### PR TITLE
[ENH] Learning Shapelet Classifier

### LINKED ISSUE #1246: [ENH] Implement Learning Shapelet algorithm for Time Series Classification
### Describe the feature or idea you want to propose

Learning shapelet is a TSC algorithm with several advantages over the Shapelet Transform algorithm. Currently, aeon doesn't have its implementation but it would be great to have one.

https://www.ismll.uni-hildesheim.de/pub/pdfs/grabocka2014e-kdd.pdf

### Describe your proposed solution

As discussed on Slack, we can wrap the TSLearn implementation for now and then improve its performance.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by itsdivya1309**
#### Reference Issues/PRs
Fixes #1246 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [ ] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).

##### For developers with write access
- [ ] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

**Comment by MatthewMiddlehurst** [comment]
Wrappers from other packages need some extra considerations. See `MrSQMClassifier` for an example.

File: aeon/benchmarking/results_loaders.py [hunk 1]
```diff
@@ -58,6 +58,11 @@
         "InceptionTimeClassifier",
         "InceptionTimeRegressor",
     },
+    "LearningShapelet": {
+        "LerningShapeletClassifier",
+        "LearningShapelet",
+        "ls",
+    },
```
Comment (MatthewMiddlehurst) [inline aeon/benchmarking/results_loaders.py:65]: We do not have any results uploaded currently, so this does not have to be included.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Some of these errors are quite strange, let me know if you need a hand.

I have updated the branch to the latest version of main since we were having some CI issues.

**Comment by MatthewMiddlehurst** [comment]
Ah we currently dont support TensorFlow for version 3.8 due to some deprecations and 3.12 because of its recency. Did not think that this would impact tslearn though?

**Comment by itsdivya1309** [comment]
> Some of these errors are quite strange, let me know if you need a hand.
> 
> I have updated the branch to the latest version of main since we were having some CI issues.

Sure, I'll try to figure it out.

**Comment by MatthewMiddlehurst** [comment]
You may have to add it as an additional dependency to the classifier for now in tags. I think SignatureClassifier has an example of a dependency that does not accept all python version

**Comment by itsdivya1309** [comment]
@MatthewMiddlehurst I think the examples are failing because we don't have any results for LearningShapelet. How can I resolve it?

**Comment by MatthewMiddlehurst** [comment]
I think you are correct, we may have to change the notebook. Any thoughts @TonyBagnall?

**Comment by TonyBagnall** [comment]
yep, I have hard coded it to the ones we have, I can run the LS after this is in and add the results. Hopefully merging once this is in will fix the above

https://github.com/aeon-toolkit/aeon/pull/1331

**Comment by TonyBagnall** [comment]
hi @itsdivya1309 if you update from main now it should be fine, thanks for doing this. I havent tried running LS for some time, it will be interesting to see how it performs

**Review by TonyBagnall — COMMENTED** [review]
sorry for how long this has taken. Things you should add 
get_test_params(
which sets the parameters to be really small, to test interface compliance. look at , for example. RDSTClassifier. 

Things that could be done in a subsequent PR: You might want to add a test function to test the locate function and also to add expected results in the test function.

**Comment by itsdivya1309** [comment]
> sorry for how long this has taken. Things you should add get_test_params( which sets the parameters to be really small, to test interface compliance. look at , for example. RDSTClassifier.
> 
> Things that could be done in a subsequent PR: You might want to add a test function to test the locate function and also to add expected results in the test function.

Thanks for looking into this, I'll add the get_test_params()

File: aeon/classification/shapelet_based/_ls.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        X_transformed = np.transpose(X, (0, 2, 1))
+    elif X.ndim == 2:
+        X_transformed = np.transpose(X)
+    return X_transformed
+
+
+class LearningShapeletClassifier(BaseClassifier):
+    """
+    Learning Shapelet.
+
```
Comment (MatthewMiddlehurst) [inline aeon/classification/shapelet_based/_ls.py:24]: Could you add a bit more documentation here above the parameters, at least mentioning that this is a wrapped for the `tslearn` package.

File: aeon/classification/shapelet_based/_ls.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    Parameters
+    ----------
+    n_shapelets_per_size: dict (default: None)
+        Dictionary giving, for each shapelet size (key),
+        the number of such shapelets to be trained (value).
+        If None, `grabocka_params_to_shapelet_size_dict` is used and the
+        size used to compute is that of the shortest time series passed at fit
+        time.a
```
Comment (MatthewMiddlehurst) [inline aeon/classification/shapelet_based/_ls.py:32]: Trailing "a" at the end, typo?

File: aeon/classification/shapelet_based/_ls.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+        Returns
+        -------
+        params : dict or list of dict, default={}
+            Parameters to create testing instances of the class.
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`.
+        """
+        return {"n_shapelets_per_size": None, "shapelet_length": 0.15}
```
Comment (MatthewMiddlehurst) [inline aeon/classification/shapelet_based/_ls.py:195]: Generally we want to set these to be fast running parameters. No need to set them if they are the default. I would change `max_iter` to something like 50.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
Looking good, very close to done I feel. Thanks for working on this. Few documentation and testing requests.

File: aeon/classification/shapelet_based/_ls.py (hunk 3 above)
Comment (itsdivya1309) [reply aeon/classification/shapelet_based/_ls.py]: Yes :)

File: aeon/classification/shapelet_based/_ls.py (hunk 4 above)
Comment (itsdivya1309) [reply aeon/classification/shapelet_based/_ls.py:206]: Well `max_iter` does have a default value of 10000, but that would indeed make it slow. So I'll set it to 50.

**Review by MatthewMiddlehurst — APPROVED** [review]
Good work, thanks for implementing my requests.

Don't mind adding to the documentation in a separate PR, up to others.

**Pull request merged**